compiler/src/abi/json_builder.rs

```diff
@@ -0,0 +1,240 @@
+    ))
+}
+
+/// Builds an ABI function element, returning None if the function is private.
+pub fn func_def<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    stmt: &'a vyp::ContractStmt<'a>,
+) -> Result<Option<Function>, CompileError> {
+    if let vyp::ContractStmt::FuncDef {
+        qual,
+        name,
+        args,
+        return_type,
+        body: _,
+    } = stmt
+    {
+        if qual.is_none() {
+            return Ok(None); // Private method.
+        }
+
+        let outputs = if let Some(return_type) = return_type {
+            vec![Output {
+                name: String::from("return_value"),
+                typ: type_desc(type_defs, &return_type.node)?,
+            }]
+        } else {
+            Vec::new()
+        };
+
+        return Ok(Some(Function {
+            name: String::from(name.node),
+            typ: FunctionType::Function,
+            inputs: args
+                .iter()
+                .map(|arg| func_def_arg(type_defs, &arg.node))
+                .collect::<Result<Vec<Input>, CompileError>>()?,
+            outputs,
+        }));
+    }
+
+    Err(CompileError::static_str(
+        "Function definition translation requires FuncDef parameter.",
+    ))
+}
+
+/// Builds a function input.
+pub fn func_def_arg<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    arg: &'a vyp::FuncDefArg<'a>,
+) -> Result<Input, CompileError> {
+    Ok(Input {
+        name: String::from(arg.name.node),
+        typ: type_desc(&type_defs, &arg.typ.node)?,
+    })
+}
+
+/// Maps the type description to an ABI-friendly type and handles custom types.
+pub fn type_desc<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    typ: &'a vyp::TypeDesc<'a>,
+) -> Result<VariableType, CompileError> {
+    if let vyp::TypeDesc::Base { base } = typ {
+        if let Some(custom_type) = type_defs.get(base) {
+            return type_desc(&HashMap::new(), custom_type);
+        }
+    }
+
+    match typ {
+        vyp::TypeDesc::Base { base: "u256" } => Ok(VariableType::Uint256),
+        _ => Err(CompileError::static_str("Unrecognized Vyper type.")),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::abi::json_builder::{Function, FunctionType, Input, VariableType, Output, type_desc, func_def};
+    use vyper_parser::parsers;
+    use vyper_parser::ast as vyp;
+    use std::collections::HashMap;
+
+    #[test]
+    fn test_type_desc_custom() {
+        let toks = vyper_parser::get_parse_tokens("CustomType").unwrap();
+        let vyp_type_desc = parsers::type_desc(&toks[..]).unwrap().1.node;
+        let mut type_defs = HashMap::new();
+        type_defs.insert("CustomType", &vyp::TypeDesc::Base { base: "u256" });
+
+        let variable_type = type_desc(&type_defs, &vyp_type_desc).unwrap();
+        assert_eq!(variable_type, VariableType::Uint256, "Incorrect type");
+    }
+
+    #[test]
+    fn test_func_def() {
+        let toks = vyper_parser::get_parse_tokens("pub def foo(x: u256) -> u256:\n   return x").unwrap();
```

Comment:
would be nice to test `pub` vs `private` fns (likely in another unit test here...)

can skip if you feel this is covered in a broader test elsewhere....